src/combinatorics/graph/basic.lean

```diff
@@ -0,0 +1,620 @@
+structure simple_multigraph (α : Type u₂) extends multigraph.{u₁ u₂} α : Type (max u₁ u₂) :=
+(hom_self (a : α) : is_empty (a ⟶ a))
+
+/-- Reinterprets a simple multigraph as a simple quiver. -/
+def simple_multigraph.to_simple_quiver (G : simple_multigraph α) : simple_quiver α := { .. G }
+
+/-- `simple_multigraph_class F α` states that `F` is a type of simple multigraphs on `α`.
+
+Extend this class when you extend `simple_multigraph`. -/
+class simple_multigraph_class (F : Type u₁) (α : out_param $ Type u₂)
+  extends multigraph_class.{u₁ u₂ u₃} F α : Type (max u₁ u₂ u₃) :=
+(hom_self (f : F) (a : α) : is_empty (a ⟶[f] a))
+
+@[priority 100] -- see Note [lower instance priority]
+instance simple_multigraph_class.to_simple_quiver_class [simple_multigraph_class F α] :
+  simple_quiver_class F α :=
+{ .. ‹simple_multigraph_class F α› }
+
+lemma is_empty.lift {α β : Type*} (h : is_empty β) (f : α → β) : is_empty α :=
+⟨λ a, is_empty_elim (f a)⟩
+
+@[priority 100] -- see Note [lower instance priority]
+instance simple_multigraph_class.to_simple_graph_class [simple_multigraph_class F α] :
+  simple_graph_class F α :=
+{ to_fun := graph.adj,
+  to_fun_symm := λ f a b h, h.map graph.hom.symm,
+  to_fun_self := graph.adj_self }
+
+instance simple_multigraph.simple_multigraph_class :
+  simple_multigraph_class (simple_multigraph α) α :=
+{ hom := λ G, G.hom,
+  hom_fun := λ G, G.hom_fun,
+  hom_fun_comp := λ G, G.hom_fun_comp,
+  hom_self := λ G, G.hom_self }
+
+instance : has_bot (simple_multigraph α) :=
+⟨{ hom := λ _ _, pempty,
+  hom_self := λ _, pempty.is_empty ,
+  hom_fun := λ _ _, id,
+  hom_fun_comp := λ _ _, rfl }⟩
+
+instance : inhabited (simple_multigraph α) := ⟨⊥⟩
+
+
+/-! #### Weight subworld -/
+
+/-- Type of `W`-weighted quivers on `α`. -/
+structure weight_quiver (α W : Type*) extends quiver α :=
+(hom_weight (a b : α) : (a ⟶ b) → W)
+
+/-- `weight_quiver_class F α W` states that `F` is a type of `W`-weighted quivers on `α`.
+
+Extend this class when you extend `weight_quiver`. -/
+class weight_quiver_class (F : Type*) (α W : out_param $ Type*) extends quiver_class F α :=
+(hom_weight (f : F) (a b : α) : (a ⟶[f] b) → W)
+
+instance weight_quiver.weight_quiver_class : weight_quiver_class (weight_quiver α W) α W :=
+{ hom := λ G, G.hom,
+  hom_weight := λ G, G.hom_weight }
+
+/-- The weight of a multiedge. -/
+protected def graph.hom.weight [weight_quiver_class F α W] {f : F} {a b : α} : (a ⟶[f] b) → W :=
+weight_quiver_class.hom_weight _ _ _
+
+instance : has_bot (weight_quiver α W) :=
+⟨{ hom := λ _ _, pempty,
+  hom_weight := λ _ _, pempty.elim }⟩
+
+instance : inhabited (weight_quiver α W) := ⟨⊥⟩
+
+
+/-- Type of `W`-weighted multigraphs on `α`. -/
+structure weight_multigraph (α : Type u₂) (W : Type u₃) extends multigraph.{u₁ u₂} α :
+  Type (max u₁ u₂ u₃) :=
+(hom_weight (a b : α) : (a ⟶ b) → W)
+(hom_weight_comm (a b : α) (e : a ⟶ b) : hom_weight _ _ (hom_fun _ _ e) = hom_weight _ _ e)
```

Comment:
Another advantage of using a single type for the edges of a multigraph : you don't have to make `a` and `b` arguments here.